File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    dp = [triangle[0]]
    for i in range(1,len(triangle)):
        arr = []
        arr.append(dp[i-1][0]+triangle[i][0])
        arr.append(dp[i-1][-1]+triangle[i][-1])
        for j in range(1,len(triangle[i])-1):
            arr.insert(j,max(dp[i-1][j-1], dp[i-1][j])+triangle[i][j])
        dp.append(arr)
    
    return max(arr)

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    let dp = Array(triangle.length).fill(0)
    
    for(let i=0;i<triangle.length;i++){
        dp[i] = Array(triangle[i].length).fill(0)
    }
    dp[0][0] = triangle[0][0]
    for(let i=1;i<triangle.length;i++){
        for(let j=0;j<triangle[i].length;j++){
            if(j===0) dp[i][j] = triangle[i][j]+dp[i-1][j]
            else if(j===triangle[i].length-1) dp[i][j] = triangle[i][j]+dp[i-1][j-1]
            else dp[i][j]=triangle[i][j] + Math.max(dp[i-1][j],dp[i-1][j-1])
        }
    }
    let max=dp[dp.length-1][0]
    for(let i=0;i<dp[dp.length-1].length;i++){
        max=Math.max(max,dp[dp.length-1][i])
    }
    return max
}